server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java

```diff
@@ -1748,10 +1719,38 @@ private void upgradeTsFileResourceCallBack(TsFileResource tsFileResource) {
           }
         }
       }
-      if (StorageEngine.getInstance().countUpgradeFiles() == 0) {
```

Comment:
So, when to stop the service?

Reply:
At the end of the `runMayThrow` method of `UpgradeTask`.